### ISSUE TITLE
Flow control on RestTemplate doesn't intercept URLs with parameters

### ISSUE THREAD

**Issue by vipweihua** [alibaba/spring-cloud-alibaba#481]
Component: Sentinel

Description:
A resource flow control (http://localhost:8002/{id}) was set, but did not take effect on RestTemplate when access the url: http://localhost:8002/123, while feign did.

STR:
1. set the rule as following:
{
    "resource": "GET:http://localhost:8001/{id}",
    "controlBehavior": 0,
    "count": 1,
    "grade": 1,
    "limitApp": "default",
    "strategy": 0
  }
2. Use RestTemplate to access http://localhost:8001/123 as following:
```
String user = restTemplate.getForEntity("http://localhost:8001/{id}", String.class, Collections.singletonMap("id", id)).getBody();
```
3. You'll see the rule doesn't take effect to intercept the url.

Investigation:
The logic of how to fetch the resource is different in SentinelProtectInterceptor and SentinelInvocationHandler. It's easy to get the accurate resource for feign in SentinelInvocationHandler because it's explicitly defined on the annotation, but it's hard to get it in SentinelProtectInterceptor for RestTemplate because the request has been evaluated before intercepted. 

It's hard, but the info is really there as you can see in the step 2 of STR.

**Comment by fangjian0423** [comment]
It seems we could not get wildcard characters in `ClientHttpRequestInterceptor`. 

Maybe we could create a class like `UrlCleaner` to clean url.

**Comment by vipweihua** [comment]
Is there any plan to fix/enhance it in 0.2.2?

**Comment by fangjian0423** [comment]
Creating UrlCleaner:

```java
public interface UrlCleaner {
    String clean(String originUrl);
}
```

Adding UrlCleaner in `SentinelProtectInterceptor` to clean url:

```
// ...
String newResource = urlCleaner.clean(hostWithPathResource);
SphU.entry(newResource, EntryType.OUT);
// ...
```

Then you could implement the path matching algorithm with `UrlCleaner`.

**Comment by fangjian0423** [comment]
> Is there any plan to fix/enhance it in 0.2.2?

Not yet, maybe next version.

**Comment by vipweihua** [comment]
> > Is there any plan to fix/enhance it in 0.2.2?
> 
> Not yet, maybe next version.

This is very important feature. Without this, it's nearly no way to use client-side flow control because the rule can't be defined for those RestTemplate has already been widely used. 
Hope it can be enhanced in 0.2.2.

**Comment by fangjian0423** [comment]
> > > Is there any plan to fix/enhance it in 0.2.2?
> > 
> > 
> > Not yet, maybe next version.
> 
> This is very important feature. Without this, it's nearly no way to use client-side flow control because the rule can't be defined for those RestTemplate has already been widely used.
> Hope it can be enhanced in 0.2.2.

Why not use Open-Feign?

**Comment by vipweihua** [comment]
> > > > Is there any plan to fix/enhance it in 0.2.2?
> > > 
> > > 
> > > Not yet, maybe next version.
> > 
> > 
> > This is very important feature. Without this, it's nearly no way to use client-side flow control because the rule can't be defined for those RestTemplate has already been widely used.
> > Hope it can be enhanced in 0.2.2.
> 
> Why not use Open-Feign?

Some URLs may be dynamic and specified by configuration after deployment. OpenFeign can only be used on the scenario that we know the APIs during coding.
